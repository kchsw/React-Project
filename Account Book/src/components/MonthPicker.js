import React, { Component } from 'react';
import PropTypes from 'prop-types'
import { padLeft, range } from '../utility'

class MonthPicker extends Component {
    constructor(props) {
        super(props)
        this.state = {
            isOpen: false,
            selectedYear: this.props.year
        }
    }
    componentDidMount() {
        document.addEventListener('click', this.handleClick, false)
    }
    componentWillUnmount() {
        document.removeEventListener('click', this.handleClick, false)
    }
    handleClick = (event) => {
        if (this.node.contains(event.target)) {
          return;
        }
        this.setState({
          isOpen: false,
        })
    }
    toggleDropdown = (e) =>{
        e.preventDefault();
        this.setState({
            isOpen: !this.state.isOpen
        })
    }
    selectYear = (e, year) => {
        e.preventDefault();
        this.setState({
            selectedYear: year
        })
    }
    selectMonth = (e, month) => {
        e.preventDefault();
        this.setState({
            isOpen: false
        })
        this.props.onChange(this.state.selectedYear, month)
    }
    render() { 
        const { year, month } = this.props
        const { isOpen, selectedYear } = this.state
        const monthRange = range(12, 1)
        const yearRange = range(9, -4).map(i => i + year)
        return (
            <div className="dropdown position-relative" ref={(ref) => {this.node = ref}}>
                <h6>选择月份</h6>
                <button className="btn btn-sm btn-secondary dropdown-toggle"
                    onClick={this.toggleDropdown}
                >
                    {`${year}年${padLeft(month)}月`}
                </button>
                {
                    isOpen && 
                    <div className="row border position-absolute" style={{background: '#fff', zIndex: 10, width: '100%'}}>
                        <div className="col border-right years-range">
                           {
                               yearRange.map((yearNumber, index) =>
                                <a key={index} 
                                    className={selectedYear === yearNumber ? "dropdown-item active" : "dropdown-item"}
                                    href="#"
                                    onClick={(e) => {this.selectYear(e, yearNumber)}}
                                >
                                   {yearNumber}年
                                </a>
                                )
                           }
                        </div>
                        <div className="col months-range">
                            {
                               monthRange.map((monthNumber, index) =>
                                <a key={index} 
                                    className={monthNumber === month ? "dropdown-item active" : "dropdown-item"}
                                    href="#"
                                    onClick={(e) => {this.selectMonth(e, monthNumber)}}
                                >
                                   {monthNumber}月
                                </a>
                                )
                            }
                        </div>
                    </div>
                }
            </div>
        );
    }
}

MonthPicker.propTypes = {
    year: PropTypes.number.isRequired,
    month: PropTypes.number.isRequired,
    onChange: PropTypes.func.isRequired
}
 
export default MonthPicker;