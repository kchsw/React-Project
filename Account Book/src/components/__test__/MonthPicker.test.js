import React from 'react';
import ReactDOM from 'react-dom';
import { mount } from 'enzyme';
import MonthPicker from '../MonthPicker'


const props = {
    year: 2019,
    month: 11,
    onChange: jest.fn()
}

let wrapper 

describe('test MonthPicker component', () => {
    beforeEach(() => {
        wrapper = mount(<MonthPicker {...props}/>)
    })
    it('snapshot', () => {
        expect(wrapper).toMatchSnapshot()
    })
    it('render correct year and month, show correct dropdown status', () => {
        const text = wrapper.find('.dropdown-toggle').first().text()
        expect(text).toEqual('2019年11月')
        expect(wrapper.state('isOpen')).toBeFalsy()
        expect(wrapper.state('selectedYear')).toEqual(props.year)
    })
    it('dropdown should show and year&month list have correct items after click button', () => {
        wrapper.find('.dropdown-toggle').simulate('click')
        expect(wrapper.state('isOpen')).toBeTruthy()
        expect(wrapper.find('.years-range .dropdown-item').length).toEqual(9)
        expect(wrapper.find('.months-range .dropdown-item').length).toEqual(12)
        expect(wrapper.find('.years-range .dropdown-item.active').text()).toEqual('2019年')
        expect(wrapper.find('.months-range .dropdown-item.active').text()).toEqual('11月')
        expect(wrapper.find('.years-range .dropdown-item').first().text()).toEqual(`${props.year - 4}年`)
    })
    it('click the year&month item should trigger right change', () => {
        wrapper.find('.dropdown-toggle').simulate('click')
        wrapper.find('.years-range .dropdown-item').first().simulate('click')
        expect(wrapper.find('.years-range .dropdown-item').first().hasClass('active')).toBeTruthy()
        expect(wrapper.state('selectedYear')).toEqual(2015)
        wrapper.find('.months-range .dropdown-item').first().simulate('click')
        expect(wrapper.state('isOpen')).toBeFalsy()
        expect(props.onChange).toHaveBeenCalledWith(2015, 1)
    })
    it('click the document should close the dropdown when the dropdown is show', () => {
        let eventMap = {}
        document.addEventListener = jest.fn((even, cb) => {
            eventMap[even] = cb
        })
        const wrapper = mount(<MonthPicker {...props}/>)
        wrapper.find('.dropdown-toggle').simulate('click')
        eventMap.click({
            target: ReactDOM.findDOMNode(wrapper.instance())
        })
        expect(wrapper.state('isOpen')).toBeTruthy()
        eventMap.click({
            target: document
        })
        expect(wrapper.state('isOpen')).toBeFalsy()
    })
})