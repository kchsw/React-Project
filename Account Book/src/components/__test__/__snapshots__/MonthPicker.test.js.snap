// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test MonthPicker component snapshot 1`] = `ReactWrapper {}`;
